Implemented Geometry.GetWidenedGeometry.

grokys opened the pull request:
## What does the pull request do?

Implements `Geometry.GetWidenedGeometry`. Similar to the WPF method [`Geometry.GetWidenedPathGeometry`](https://learn.microsoft.com/en-us/dotnet/api/system.windows.media.geometry.getwidenedpathgeometry?view=windowsdesktop-7.0) except that it doesn't return a `PathGeometry` and there is no overload with tolerance settings.

grokys committed:
Implemented Geometry.GetWidenedGeometry.

grokys commented on src/Avalonia.Base/Media/ImmutableGeometry.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Avalonia.Media;
+
+internal class ImmutableGeometry : Geometry
```
Introduced this to store the returned widened geometry. Because the geometry can't change once calculated, it's "immutable" but this naming may be confusing as we have other immutable public drawing primitives in `Avalonia.Media.Immutable`?

kekekeks replied:
`PlatformGeometry`?

kekekeks commented on src/Skia/Avalonia.Skia/GeometryImpl.cs:
```diff
@@ -191,6 +208,10 @@ public void UpdateIfNeeded(SKPath? strokePath, IPen? pen)
                 paint.StrokeJoin = join.ToSKStrokeJoin();
                 paint.StrokeMiter = (float)miterLimit;
+
+                if (DrawingContextHelper.TryCreateDashEffect(pen, out var dashEffect))
```
Since it's used for caching, we need to record/compare the dash effect too.

kekekeks replied:
It's used in `StrokeContains` which can be called kinda frequently.

grokys replied:
We previously weren't using the dash effect for `StrokeContains`. Should we conditionally enable it?

kekekeks replied:
_Previously_ we've only used stroke thickness. It was updated two weeks ago to use some other pen properties so render bounds are calculated properly.

kekekeks replied:
Might be worth to extract the code that creates the Path into a separate method

grokys replied:
Done

kekekeks commented on src/Skia/Avalonia.Skia/GeometryImpl.cs:
```diff
@@ -75,6 +76,22 @@ public Rect GetRenderBounds(IPen? pen)
 
+        public IGeometryImpl GetWidenedGeometry(IPen pen)
+        {
+            var cache = new PathCache();
```
`using var`

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0038940-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

grokys committed:
Dispose of the PathCache.

grokys committed:
Only include dash style when widening bounds.

Otherwise we'd be changing the behavior of an existing API (`StrokeContains`) which should be discussed first.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0038954-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

grokys committed:
Refactor creating a stroked path.

- Move the logic into `SKPathHelper.CreateStrokedPath`; called from `PathHelper` and `GetWidenedGeometry`
- Use a pen hash code to check if we're up-to-date in `PathHelper` and include the dash style in that

maxkatz6 commented on src/Skia/Avalonia.Skia/Helpers/PenHelper.cs:
```diff
@@ -0,0 +1,38 @@
+        if (pen is null)
+            return 0;
+
+        var hash = new HashCode();
```
I don't think it's available in netstandard2.0.

grokys replied:
Damn, didn't mean to push that commit yet! Yeah I was going to include a polyfil for that to Avalonia.Base. Maybe based on https://github.com/mono/SkiaSharp/blob/main/binding/Binding.Shared/HashCode.cs? Good idea or not?

grokys replied:
Added.

grokys committed:
Added polyfil of System.HashCode.

kekekeks commented on src/Skia/Avalonia.Skia/Helpers/SKPathHelper.cs:
```diff
@@ -0,0 +1,67 @@
+        paint.StrokeJoin = pen.LineJoin.ToSKStrokeJoin();
+        paint.StrokeMiter = (float)pen.MiterLimit;
+
+        if (DrawingContextHelper.TryCreateDashEffect(pen, out var dashEffect))
```
Path effect is disposable

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0039027-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

grokys committed:
Dispose path effect.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0039034-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]
